### ISSUE TITLE
socket is not active if session is out of scope

### ISSUE THREAD

**Issue by inv2004** [blabaere/scaproust#7]
Hello,

Please find part of my code, it is not full, but I hope problem is clear from it: https://gist.github.com/anonymous/1087c06617dccd8b8f1d9c13e3da7699

If session goes out of scope - it do something with socket that it does not work. It is necessary to prolongate session lifetime then socket is ok.

Is it scaproust lib bug?

Thank you,

**Comment by blabaere** [comment]
Hello,

Yes you need the session to be alive for anything to work. Once the session goes out of scope, the "backend" system is terminated and anything created by the session will also stop working.

Regards,
Benoît

**Comment by inv2004** [comment]
Is it possible to make it owned by the Socket? Because now it is necessary to hold pair or structure to keep it.

Regards,

**Comment by blabaere** [comment]
It is not possible to make the session owned by the socket, at least with the current design. 
First, the session is meant to create several sockets, and because the socket and device creation functions require a mutable session.

I guess a solution for your use case would be, like you mentioned, to make a new struct, owning the session and the socket, and put all the required functions on that new struct type.

I'm aware this is less convenient than the current nanomsg design for example but I have not been able to find a design addressing all the concerns for the moment.

Regards,
Benoît

**Comment by eberkund** [comment]
I came across this because I was curious about the session as well. As far as I know unlike ZeroMQ, nanomsg library doesn't use a session to store state. See for example: https://github.com/go-mangos/mangos/blob/master/examples/reqrep/reqrep.go

I guess this is a bit of a deviation in the design?

**Comment by blabaere** [comment]
Scaproust has a background worker thread for all I/O and since Rust requires every memory region to have a unique owner, I had to create one. I chose to introduce a session and to make the end-user responsible for creating and destroying it because it was much easier than introduce a shared mutable static variable.

If I had to choose now, I would probably design it differently. For example I could require the end-user to deal with creating the tokio event-loop, and the need for a session would be much less obvious.
